prevent null client unregister

dpatekar opened the pull request:
When using TLS like this:
```
var tlsServer = tls.createServer(options, function(s){  
  console.log('secure connection established');
  aedes.handle(s);
});
tlsServer.listen(config.ports.tcptls);
```
Sometimes happens that client does not connect completely (e.g. using some ssl scan tool like https://github.com/rbsec/sslscan) and **end-of-stream** is firing a callback which is **client.close()**. In that situation, client is not connected to the broker. That causes incorrect decrement of connected clients in broker.unregisterClient()

mcollina commented:
Can you please add a unit test for this?
Il giorno mer 17 feb 2016 alle 19:56 dpatekar <notifications@github.com> ha
scritto:

> When using TLS like this:
> var tlsServer = tls.createServer(options, function(s){
> console.log('secure connection established');
> aedes.handle(s);
> });
> tlsServer.listen(config.ports.tcptls);
>
> Sometimes happens that client does not connect completely (e.g. using some
> ssl scan tool like https://github.com/rbsec/sslscan) and *end-of-stream*
> is firing a callback which is *client.close()*. In situation, client is
> not connected to the broker. That causes incorrect decrement of connected
> clients in broker.unregisterClient()
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mcollina/aedes/pull/32
> Commit Summary
>
>    - Merge pull request #1 from mcollina/master
>    - Prevent null client unregister.
>
> File Changes
>
>    - *M* lib/client.js
>    <https://github.com/mcollina/aedes/pull/32/files#diff-0> (4)
>
> Patch Links:
>
>    - https://github.com/mcollina/aedes/pull/32.patch
>    - https://github.com/mcollina/aedes/pull/32.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mcollina/aedes/pull/32>.
>

dpatekar commented:
I have added a unit test in the existing one "authenticate unsuccessfully a client with username and password" because that situation reproduces the issue. Just check for connectedClients count on end-of-stream.

In addition to this, I have added a unit test and solution to the situation when a client subscribes to some topic with QoS 1 and then publishes a packet to the same topic with QoS 0. Persistence module is throwing an error in that situation because there is no downgrade to QoS 0 and according to the MQTT specs it should be because subscribing QoS should define the maximum QoS for that topic. What do you think about this?

mcollina commented:
That's great and awesome! Thanks so much for spotting the QoS downgrading!
